Updates Deployment doc (#181)
tektoncd/hub: docs/DEPLOYMENT.md

# Deployment

- [Prerequisites](#prerequisites)
- [Deploy API Service](#deploy-api-service)
  - [Update API Secret](#update-the-api-secret)
  - [Update API Config Map](#update-api-config-map)
  - [Prepare API and Db Migration release yaml](#Prepare-the-api-and-db-migration-release-yaml)
  - [Setup Ingress/Route](#setup-route/ingress)
- [Deploy UI](#deploy-ui)
  - [Build and Publish Image](#build-and-publish-image)
  - [Update the UI ConfigMap](#update-the-ui-configmap)
  - [Setup Ingress/Route](#setup-route/ingress)
- [Add resources in DB](#add-resources-in-db)
- [Setup Catalog Refresh CronJob](#setup-catalog-refresh-cronjob)

> - [Adding New Users in Config](#adding-new-users-in-config)

Reviewer: Are we missing to add a heading and description for adding a jwt token which does not gets expired ??

Author: that is in the catalog refresh part

Navigate to the `config/00-config/` and edit `31-api-secret.yaml` . Set GitHub `oauth` client id and client secret.

> To create a GitHub OAuth follow the steps given [here][oauth-steps]. For now add any URL in place for `Homepage URL` and `Authorization callback URL` for eg. `http://localhost:5000`. We will update this URL once we deploy the UI pod.

Reviewer: Could you please create links for all the headings as we have in the development docs ?

cd api
```

### Update the API Secret

Navigate to the `config/00-config/` and edit `31-api-secret.yaml` . Set GitHub `oauth` client id and client secret.

> To create a GitHub OAuth follow the steps given [here][oauth-steps]. For now add any URL in place for `Homepage URL` and `Authorization callback URL` for eg. `http://localhost:5000`. We will update this URL once we deploy the UI pod.

Reviewer 2: ```suggestion
To create a GitHub OAuth follow the steps given [here][oauth-steps]. For now add any URL in place of `Homepage URL` and `Authorization callback URL` for eg. `http://localhost:5000`. We will update this URL once we deploy the UI pod.
```

  REFRESH_JWT_EXPIRES_IN: time such as 15m
```

#### Config File

> If you want to change the config file passed to hub, you can edit the `30-api-configmap.yaml` and change the URL to the file. By default it is pointing to [config.yaml][config.yaml] in Hub which has Application Data.

Reviewer: Could you please provide the link to `30-api-configmap.yaml` and a code block of where the user needs to edit

Author: sure

> In this file we add users who have additional scopes which can be used to refresh a catalog, refresh config file after changes, create an agent token which can be used in cron job to refresh the catalog after an interval.

Reviewer: Refresh config file after changes ? :thinking:

Author: yeah after adding new users

> All users have default scopes they get after login which allows them to rate the resources.

Reviewer: Here you have mentioned that it allows to rate the resources, instead can we say or mention the default scopes are `rating:read` and `rating:write`

  REFRESH_JWT_EXPIRES_IN: time such as 15m
```

### Update API Config Map

> If you want to change the config file passed to hub, you can edit the `[30-api-configmap.yaml](https://github.com/tektoncd/hub/blob/master/api/config/00-config/30-api-configmap.yaml)` and change the URL to the file. By default it is pointing to [config.yaml][config.yaml] in Hub which has Application Data.

Reviewer: I think it has not created a link here

Apply the release yaml

```bash
kubectl apply -f api.yaml
```

The command above will create the deployment for Database and API and a job for DB-Migration.

The db-migration job will create all tables in the db.

```bash

> $ kubectl get pods

Reviewer 2: Since we are creating this in a namespace, so can we add `-n <namespace>` in this command?

For `OpenShift`:-

```bash
curl -k -X GET -I $(oc get routes api --template='https://{{ .spec.host }}/v1/categories')
```

> NOTE: At this moment, there are no resources in the db. Only the tags and categories from hub config are added in the db.

Reviewer: After creating the link, can you add the link here with a line saying, for adding the resources follow the catalog refresh steps, wdyt ??

Reviewer: I know we have added to deploy the ui first but just a suggestion

docker build -t <image> . && docker push <image>
```

> eg. `quay.io/<username>/ui`

Reviewer: Why do we need this here ?? :thinking:

Author: its an example

Reviewer: And also mention if one pushes the image to quay, he needs to make that image public

    kubectl apply -f config/99-post-deploy/33-api-ingress.yaml
    ```

#### Verify if api route is accessible

For `OpenShift`:-

```bash

> curl -k -X GET -I $(oc get routes api --template='https://{{ .spec.host }}/v1/categories')

Reviewer 2: same as above, add namespace to the command?

docker build -t <image> . && docker push <image>
```

eg. `quay.io/<username>/ui`

Make sure your image is public after pushing to a registry.

#### Update the deployment image

> Update `config/11-deployement` to use the image built above

Reviewer 2: ```suggestion
Update `config/11-deployment` to use the image built above
```

docker build -t <image> . && docker push <image>
```

eg. `quay.io/<username>/ui`

Make sure your image is public after pushing to a registry.

#### Update the deployment image

> Update `config/11-deployment` to use the image built above

Reviewer: Should we add a code block and show where to update the image

  name: ui
  namespace: tekton-hub
data:

>   API_URL: API URL   <<< update this by api route

Reviewer 2: Although it is very obvious to get the api route using `oc get route` command. But to be on a safer side can we also add this command here as an comment?

  GH_CLIENT_ID: GH OAuth Client ID   <<< update this
```

#### Apply the manifests

```bash
kubectl apply -f config/
```

This will create the deployment and service for UI.

#### Ensure pods are up and running

```bash

> kubectl get pods

Reviewer 2: same as above, can we also add namespace flag here as well and in all the commands?
or also we can change the context to the desired namespace, so that we don't have to pass namespace in every command.

Open: `oc get routes ui --template='https://{{ .spec.host }} '`

> Note: In UI you won't be able to resources as there are not any in the db. You can see Categories on the left side.

Reviewer: Can we change this to 
```suggestion
Note: In UI you won't be able to resources as there are not any in the db, but you should see Categories on the left side.
```

Update your GitHub OAuth created with the UI route in place of `Homepage URL` and `Authorization callback URL`.

### Add resources in DB

> Note: Make sure you have added your name in the scopes present [config.yaml][config.yaml]. You will need additional scope i.e. catalog-refresh to add resources.

Now, follow the below steps

1. Login through the UI. Click on `Login` on right corner and then `Sign in with GitHub`.

>     This will add you in db, but yet you have only the default scopes i.e. `rating:read` and `rating:write` which you can check with your jwt. On UI, in right corner there is an option to `Copy Hub Token` and paste in any jwt decoder. for eg. https://jwt.io/

Reviewer 3: why do we need to paste in the JWT decoder?

Author: just for fun :D....no reason so that user can check scopes

    
2. We save the checksum of config file in db to avoid reading the config file again and again if the api pod is deleted due to any reason. The config file is read by API pod before starting the server. 

3. To make API pod read config we need to delete the entry in db. Exec into the database pod
   ```bash
   kubectl exec -it <db-pod-name> bash
   ```
4. Open PostgreSQL terminal by executing the command

   ```bash
   psql -U postgres -d hub
   ```

   You'll see the terminal changed to

   ```bash
   hub=#
   ```
5. Execute the SQL query

   ```sql
   hub=# delete from configs;
   ```

6. Quit the database pod by first executing

   ```sql
   hub=# \q
   ```

   and then `exit`.

7. Delete the API pod by executing the following command
   ```bash
   oc delete pod <pod-name>
   ```
   This will delete the previously created pod and spin a new pod which will add the scopes in the DB.
   
8. Once the pod is running again, Logout from Hub UI and login again.

> 9. Now, if you copy your token from UI and check in jwt decoder, you will have additional scopes.

Author: that is added at the start of add resources

> 10. To add resources, you need to a POST api call passing you jwt token in Header.

Reviewer: ```suggestion
10. To add resources, you need to a POST api call passing your jwt token with the addtional scopes in Header.
```

```
curl -X POST -H "Authorization: <access-token>" \
    <api-route>/catalog/refresh 
```

> Replace you Hub token copied from Hub UI with `<access-token>` and replace `<api-route>` with you api service route/ingress.

Reviewer: Can we improve this, seems to be a little confusing

    ```
    curl -X PUT --header "Content-Type: application/json" \
        -H "Authorization: <access-token>" \
        --data '{"name":"catalog-refresh-agent","scopes": ["catalog:refresh"]}' \
        <api-route>/system/user/agent
    ```
    Replace `<access-token>` with your JWT, you can create an agent with any required scopes. In this case we need the agent with `catalog:refresh`. You can give any name to agent.

    This will gives an output as:
    ```
    {"token":"agent jwt token"}
    ```
    A token will be returned with the requested scopes. You can check using the jwt decoder. Use this token for the catalog refresh cron job.

4. Navigate to `api/config/99-post-deploy` and edit `34-catalog-refresh-secret.yaml`. Set the Hub Token

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: catalog-refresh
     namespace: tekton-hub
   type: Opaque
   stringData:
     HUB_TOKEN: hub token
   ```
    Set the token created in last step as Hub token.

4. Make sure you are in `api` directory and run the below command. 

    ```

>     kubectl apply -f config/99-post-deploy/34-catalog-refresh-secret.yaml 

Reviewer: Can't we just run 
```suggestion
    kubectl apply -f config/99-post-deploy/
```

Author: it has an ingress file also

    ```
    curl -X PUT --header "Content-Type: application/json" \
        -H "Authorization: <access-token>" \
        --data '{"name":"catalog-refresh-agent","scopes": ["catalog:refresh"]}' \
        <api-route>/system/user/agent
    ```
    Replace `<access-token>` with your JWT, you can create an agent with any required scopes. In this case we need the agent with `catalog:refresh`. You can give any name to agent.

    This will gives an output as:
    ```
    {"token":"agent jwt token"}
    ```
    A token will be returned with the requested scopes. You can check using the jwt decoder. Use this token for the catalog refresh cron job.

4. Navigate to `api/config/99-post-deploy` and edit `34-catalog-refresh-secret.yaml`. Set the Hub Token

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: catalog-refresh
     namespace: tekton-hub
   type: Opaque
   stringData:
     HUB_TOKEN: hub token
   ```
    Set the token created in last step as Hub token.

4. Make sure you are in `api` directory and run the below command. 

    ```

>     kubectl apply -f config/99-post-deploy/34-catalog-refresh-secret.yaml 

Reviewer 2: ```suggestion
    kubectl apply -f config/99-post-deploy/34-catalog-refresh-secret.yaml \
```

    kubectl apply -f config/post-deploy
    ```

#### Verify if UI route is accessible

If on openshift verify if the ui route is accessible

Open: `oc get routes ui --template='https://{{ .spec.host }} '`

> Note: In UI you won't be able to resources as there are not any in the db but you should see Categories on the left side.

Reviewer: ```suggestion
Note: In UI you won't be able see the resources on home page as there are not any in the db but you should see Categories on the left side.
```

    
2. We save the checksum of config file in db to avoid reading the config file again and again if the api pod is deleted due to any reason. The config file is read by API pod before starting the server. 

3. To make API pod read config we need to delete the entry in db. Exec into the database pod
   ```bash
   kubectl exec -it <db-pod-name> bash
   ```
4. Open PostgreSQL terminal by executing the command

   ```bash
   psql -U postgres -d hub
   ```

   You'll see the terminal changed to

   ```bash
   hub=#
   ```
5. Execute the SQL query

   ```sql
   hub=# delete from configs;
   ```

6. Quit the database pod by first executing

   ```sql
   hub=# \q
   ```

   and then `exit`.

7. Delete the API pod by executing the following command
   ```bash
   oc delete pod <pod-name>
   ```
   This will delete the previously created pod and spin a new pod which will add the scopes in the DB.
   
8. Once the pod is running again, Logout from Hub UI and login again.

9. Now, if you copy your token from UI and check in jwt decoder, you will have additional scopes.

> 10. To add resources, you need to a POST api call passing your jwt token with the addtional scopes in Header.

Reviewer: ```suggestion
10. To add resources, you need to make a POST api call passing your jwt token with the addtional scopes in Header.
```

   ```yaml
   apiVersion: v1
   kind: Secret
   metadata:
     name: catalog-refresh
     namespace: tekton-hub
   type: Opaque
   stringData:
     HUB_TOKEN: hub token
   ```
    Set the token created in last step as Hub token.

4. Make sure you are in `api` directory and run the below command. 

    ```
    kubectl apply -f config/99-post-deploy/34-catalog-refresh-secret.yaml \ 
                -f config/99-post-deploy/35-catalog-refresh-cronjob.yaml 
    ```
    The cron job is configured to run every 30 min, you can change the interval in `config/99-post-deploy/35-catalog-refresh-cronjob.yaml`.
    
## Adding New Users in Config

By default when any user login for the first time, they will have only default scope even if they are added in config.yaml. To get additinal scopes, make sure the user has logged in once.

Now, we need to refresh the config. To do that do a POST API call.

```
curl -X POST -H "Authorization: <access-token>" \
    --header "Content-Type: application/json" \
    --data '{"force": true} \
  <api-route>/system/config/refresh
```
Replace `<access-token>` with your JWT token. you must have `config-refresh` scope to call this API. 

## Troubleshooting

> Case 1: UI is not showing resources but catalog refresh is successful

Reviewer 3: nit:
```suggestion
Case 1: UI is not showing resources but catalog refresh is successful.
```

  If you have insecure connection the UI might not be able to hit the API.

>   So, hit the API URL in your browser `https://<api-url>` once you get the response as `status: ok`, refresh you UI.

Reviewer 3: same as previous

> Case 2: UI is not showing resources but catalog refresh is successful (Console Error)

Reviewer 3: ```suggestion
Case 2: UI is not showing resources but catalog refresh is successful (Console Error).
```

  Verify catalog refresh is successful by hitting the `/v1/resources` API using the API URL.
  If is returning resources than check for console errors on UI. 
  Right click -> Inspect or (Ctrl + Shift + I) (For Chrome)
  Then click on console tab.
  If there is cors error then check the UI config map and the URL you have added.

>   The URL should be for example `https://api.hub.tekton.dev` and not `https://api.hub.tekton.dev/` there shouldn't be `/` at the end.

Reviewer 3: Can you please add these in points as in preview mode they are being shown as one paragraph?
